### DISCUSSION TITLE
Generating an animation of Hole Etching Simulation from VTP File

### DISCUSSION THREAD

**Discussion by jais001** [ViennaTools/ViennaPS#91] [Q&A] [answered]
Hey ViennaPS Team,

I’ve been working on a simulation involving hole etching using ViennaPS, and I’ve successfully generated a series of simulations. Now, I would like to create an animated GIF that visualizes the etching process as a continuous, time-evolving sequence. Essentially, I want the simulation to be compiled into a GIF that clearly illustrates the progression of the etching process.

I have a few questions regarding this:

1. Does ViennaPS offer any built-in functionality(python scripts) for generating such a GIF directly from the VTP file produced during the simulation?
2. If not, could you recommend any specific Python packages that can achieve this, if there are any?

Any guidance or suggestions on how to approach this would be greatly appreciated. Thanks

**Comment by tobre1** [comment, marked as answer]
Hi there!

ViennaPS doesn’t provide built-in visualization tools, so I again recommend using ParaView for this purpose. ParaView also offers a Python package that allows you to script visualizations in [Python](https://www.paraview.org/Wiki/PvPython_and_PvBatch).

To create an animation for the SF6O2 hole etching example, you can follow these steps:

1. **Set Up Your Model:** Configure your model with the desired parameters and create a `viennaps::Process` object. Set the process duration to the desired time steps for generating output files.
2. **Run the Process and Save Outputs:** Use a loop to run the process step-by-step, saving the surface mesh after each step with the `saveSurfaceMesh` function. Append an index to each filename so that ParaView can automatically recognize them as a time series.
3. **Create the Animation in ParaView:** In the ParaView GUI, open the files as a group, and use the VCR controls in the toolbar to play the animation. You can also save the animation via `File -> Save Animation`, where you can adjust the framerate. The same can be done using the ParaView Python package (see the [documentation](https://docs.paraview.org/en/latest/UsersGuide/dataIngestion.html#id1)).

Here's a code snippet illustrating how the ViennaPS Python script should look:

```python
# Initialize the geometry
geometry = vps.Domain()

# Set up the model with your parameters
model = vps.SF6O2Etching()

# Create the Process object
process = vps.Process(geometry, model)
process.setProcessDuration(timeSteps)  # Time between intermediate outputs

time = 0
counter = 0
while time < totalProcessDuration:

    process.apply()  # Run the process for a time step
    geometry.saveSurfaceMesh("output/surfaceMesh_" + str(counter) + ".vtp")  # Save the surface mesh

    time += process.getProcessDuration()
    counter += 1
```

### Notes:
1. **Avoid Reinitializing the Process:** It’s efficient to avoid creating a new `Process` object in each iteration, as it will reinitialize the coverage each time, which can be time-consuming.
2. **Intermediate Output Option:** You can also enable intermediate outputs by setting the logging level to `vps.Logger.setLogLevel(vps.LogLevel.INTERMEDIATE)`. This will save a point cloud of the geometry with relevant data at each advection time step during the `process.apply()` call. While this method doesn’t require looping, point clouds may not always be ideal for visualization purposes.

**Comment by jais001** [comment]
@tobre1, Thanks for your input. I will try this out.
